src/presence/__tests__/heartbeat-test.js

```diff
@@ -0,0 +1,243 @@
+// @flow strict-local
+import LolexModule from 'lolex';
+import Heartbeat from '../heartbeat';
+
+/*
+ * At present (Jest v24.9.0), Jest does not override Date.now() when using a
+ * fake timer implementation. This means that any timer-based code relying on
+ * Date.now() for throttling (etc.) will be very confused.
+ *
+ * The good news is that Jest is very close to using Lolex internally -- see,
+ * e.g., https://github.com/facebook/jest/pull/7776 -- at which point that
+ * behavior will be available to us via Jest. The bad news, alas, is that it's
+ * not there yet.
+ *
+ * For now, we borrow slices of Jest's planned Lolex-based timer implementation.
+ */
+
+/**
+ * A Lolex-backed implementation of certain relevant Jest functions.
+ *
+ * Carved from the more-complete, not-yet-NPM-available implementation at:
+ * https://github.com/facebook/jest/blob/9279a3a97/packages/jest-fake-timers/src/FakeTimersLolex.ts
+ */
+class Lolex {
+  /* eslint-disable no-underscore-dangle */
+
+  /** The installed Lolex clock object. (Name also taken from Jest's
+      implementation, for simplicity's sake. */
+  _clock;
+  constructor() {
+    this._clock = LolexModule.install();
+  }
+
+  clearAllTimers(): void {
+    this._clock.reset();
+  }
+
+  getTimerCount(): number {
+    return this._clock.countTimers();
+  }
+
+  runOnlyPendingTimers(): void {
+    this._clock.runToLast();
+  }
+
+  advanceTimersByTime(msToRun: number): void {
+    this._clock.tick(msToRun);
+  }
+
+  dispose(): void {
+    this._clock.uninstall();
+  }
+}
+
+let lolex: Lolex;
+
+// type alias for Jest callback functions of type (boolean) => void
+type CallbackType = JestMockFn<$ReadOnlyArray<void>, void>;
+
+describe('Heartbeat', () => {
+  // arbitrarily, one full hour between heartbeats
+  const HEARTBEAT_TIME = 60 * 60 * 1000;
+
+  // before running tests: set up fake timer API
+  beforeAll(() => {
+    // jest.useFakeTimers();
+    lolex = new Lolex();
+  });
+
+  afterAll(() => {
+    // jest.useRealTimers();
+    lolex.dispose();
+  });
+
+  // before each test: reset fake-timers state
+  beforeEach(() => {
+    lolex.clearAllTimers();
+  });
+
+  // after each test: confirm that all timers have been stopped
+  afterEach(() => {
+    expect(lolex.getTimerCount()).toBe(0);
+  });
+
+  // convenience function: create a new Heartbeat with its associated callback
+  const setup = (): {|
+    callback: CallbackType,
+    heartbeat: Heartbeat,
+  |} => {
+    const callback: CallbackType = jest.fn();
+    const heartbeat = new Heartbeat(callback, HEARTBEAT_TIME);
+    return { callback, heartbeat };
+  };
+
+  // Heartbeat erases its callback type (and it should probably be private
+  // anyway), so the `expectRunning` and `expectNotRunning` helpers take the
+  // callback mock as a separate well-typed argument.
```

Comment:
Huh, I'm puzzled -- what do you mean by saying it erases its callback type? Here's the definition in the early part of the branch:
```
class Heartbeat {
  intervalId: IntervalID | null = null;

  callback: (state: boolean) => void;
  milliseconds: number;
```

and a later commit does `s/state: boolean//`.

It seems like the type is right there. What am I missing?

Reply:
That's not the full type, though; it's just the maximal supertype. The "underlying type" of the `callback` object here is `CallbackType`.

*Not* erasing the type would involve giving `Heartbeat` a generic type parameter `T: (boolean) => void` (later `() => void`). This isn't difficult, of course; but it's also pointless unless `callback` is part of `Heartbeat`'s public interface. (Which it really shouldn't be!)